File: src/Tests.fs/TankTests.fs
﻿module TLK.LoveDeathFSharp.Tests.Tank

open NUnit.Framework
open TLK.LoveDeathFSharp.Kernel

let shouldThrowException action = Assert.That(new TestDelegate(action), Throws.Exception)

let shoot (tank : Tank) = tank.Shoot

[<Test>]
let ``tank should not shoot with wolf inside`` () =
    { new IShapeshifter with member x.Shape = Shape.Wolf }
    |> Tank
    |> shoot
    |> shouldThrowException

File: src/Tests.cs/TankTests.cs
using System;
using FluentAssertions;
using Moq;
using NUnit.Framework;
using TLK.LoveDeathFSharp.Kernel;

namespace TLK.LoveDeathFSharp.Tests
{
    internal class TankTests
    {
        [Test]
        public void TankShouldNotShootWithWolfInside()
        {
            // arrange
            var wolf = Mock.Of<IShapeshifter>(x => x.Shape == Shape.Wolf);
            var tank = new Tank(wolf);

            // act
            Action act = () => tank.Shoot();

            // assert
            act.Should().Throw<Exception>();
        }
    }
}
